### PR TITLE
Email confirmation for mobile app

### PR DESCRIPTION
**Pull request by Iwasherd**
Create universal links for web, android and ios devices issue #740

### CODE REVIEW

File: modules/core/client-react-native/App.tsx [hunk 1]
```diff
@@ -1,9 +1,8 @@
 import React from 'react';
 import { ApolloProvider } from 'react-apollo';
 import { createStore, combineReducers } from 'redux';
-import { Provider } from 'react-redux';
 import url from 'url';
-
+import { Provider } from 'react-redux';
```
Comment (alekseyzadorozhniy) [inline modules/core/client-react-native/App.tsx]: Unnecessary changes

File: modules/user/client-react/components/LoginView.native.jsx [hunk 2]
```diff
@@ -20,18 +20,20 @@ class LoginView extends React.PureComponent {
...
-    const decodedData = JSON.parse(decodeURI(data));
-    const { client } = this.props;
-    if (decodedData.tokens) {
-      await setItem('accessToken', decodedData.tokens.accessToken);
-      await setItem('refreshToken', decodedData.tokens.refreshToken);
-    }
-    await client.query({ query: CURRENT_USER_QUERY });
+    if (url.match(/data=([^#]+)/)) {
+      // Extract stringified user string out of the URL
+      const [, data] = url.match(/data=([^#]+)/);
```
Comment (alekseyzadorozhniy) [inline modules/user/client-react/components/LoginView.native.jsx]: You can specify the regex as a constant to avoid duplication.

File: modules/user/client-react/components/ActivateUser.native.jsx [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+  render() {
+    const { t } = this.props;
+
+    return (
+      <View style={styles.container}>
+        <ActivityIndicator size="large" color="#0000ff" />
+        <Text style={styles.text}>
+          {t('activateUser')}
+          ...
```
Comment (alekseyzadorozhniy) [inline modules/user/client-react/components/ActivateUser.native.jsx]: Why are these dots here, and not in a file with localization?

File: modules/user/client-react/containers/ActivateUser.jsx [hunk 4]
```diff
@@ -0,0 +1,65 @@
+import React from 'react';
+import { graphql } from 'react-apollo';
+
+import ActivateUser from '../components/ActivateUser';
+import ACTIVATE_USER from '../graphql/ActivateUser.graphql';
+
+class Activate extends React.Component {
+  constructor(props) {
```
Comment (alekseyzadorozhniy) [inline modules/user/client-react/containers/ActivateUser.jsx]: Redundant code

File: modules/user/client-react/containers/ActivateUser.jsx [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+  componentDidMount() {
+    this.activate();
+  }
+
+  activate = () => {
+    try {
+      const { history, navigation } = this.props;
+      if (history) {
+        const token = this.props.match.params.token;
```
Comment (alekseyzadorozhniy) [inline modules/user/client-react/containers/ActivateUser.jsx]: You can use destructuring instead

File: modules/user/client-react/containers/ActivateUser.jsx [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        this.props.activateUser(token, history);
+      } else {
+        const { url } = navigation.state.params;
+        const [, token] = url.split('/confirmation/');
+        this.props.activateUser(token, navigation);
```
Comment (alekseyzadorozhniy) [inline modules/user/client-react/containers/ActivateUser.jsx]: You can get `activateUser` via destructuring

File: modules/user/client-react/containers/ActivateUser.jsx [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+export default graphql(ACTIVATE_USER, {
+  props: ({ ownProps: { history, navigation }, mutate }) => ({
+    activateUser: async token => {
+      try {
+        const {
+          data: { activateUser }
+        } = await mutate({
+          variables: { token }
+        });
+        if (activateUser.success) {
```
Comment (alekseyzadorozhniy) [inline modules/user/client-react/containers/ActivateUser.jsx]: Can you simplify this block of code?

File: modules/user/client-react/index.native.jsx [hunk 8]
```diff
@@ -135,6 +136,15 @@ ProfileScreen.propTypes = {
   navigation: PropTypes.object
 };
 
+class ActivateUserScreen extends React.Component {
```
Comment (alekseyzadorozhniy) [inline modules/user/client-react/index.native.jsx]: Can you use Function component instead of Class component?

File: modules/user/client-react/components/ActivateUser.web.jsx [hunk 9]
```diff
@@ -0,0 +1,24 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { translate } from '@module/i18n-client-react';
+import { LayoutCenter } from '@module/look-client-react';
+
+class ActivateUser extends React.PureComponent {
```
Comment (alekseyzadorozhniy) [inline modules/user/client-react/components/ActivateUser.web.jsx]: Can you use Function component instead of Class component?

File: modules/core/client-react-native/App.tsx (hunk 1 above)
Comment (Iwasherd) [reply modules/core/client-react-native/App.tsx]: Done

File: modules/user/client-react/components/LoginView.native.jsx (hunk 2 above)
Comment (Iwasherd) [reply modules/user/client-react/components/LoginView.native.jsx]: Done

File: modules/user/client-react/components/RegisterView.native.jsx [hunk 10]
```diff
@@ -1,14 +1,35 @@
...
+    Linking.addEventListener('url', this.hundlerUrl);
+    if (Platform.OS === 'ios') {
+      const url = await Linking.getInitialURL();
+      if (url.includes('/confirmation/')) {
+        this.redirectOnActivateUser(url);
+      }
+    }
+  }
+
+  hundlerUrl = ({ url }) => {
```
[the remark this answers is not in the document]
Comment (Iwasherd) [reply modules/user/client-react/components/RegisterView.native.jsx]: Done

File: modules/user/client-react/components/ActivateUser.native.jsx (hunk 3 above)
Comment (Iwasherd) [reply modules/user/client-react/components/ActivateUser.native.jsx]: Done

File: modules/user/client-react/containers/ActivateUser.jsx (hunk 4 above)
Comment (Iwasherd) [reply modules/user/client-react/containers/ActivateUser.jsx]: Done

File: modules/user/client-react/containers/ActivateUser.jsx (hunk 5 above)
Comment (Iwasherd) [reply modules/user/client-react/containers/ActivateUser.jsx]: Done

File: modules/user/client-react/containers/ActivateUser.jsx (hunk 6 above)
Comment (Iwasherd) [reply modules/user/client-react/containers/ActivateUser.jsx]: Done

File: modules/user/client-react/index.native.jsx (hunk 8 above)
Comment (Iwasherd) [reply modules/user/client-react/index.native.jsx]: Done

File: modules/user/client-react/components/ActivateUser.web.jsx (hunk 9 above)
Comment (Iwasherd) [reply modules/user/client-react/components/ActivateUser.web.jsx]: Done

File: modules/user/client-react/containers/ActivateUser.jsx [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    activateUser: async token => {
+      try {
+        const {
+          data: {
+            activateUser: { success }
+          }
+        } = await mutate({
+          variables: { token }
+        });
+        const mobileActions = success ? navigation.navigate('Login') : navigation.navigate('Counter');
```
Comment (alekseyzadorozhniy) [inline modules/user/client-react/containers/ActivateUser.jsx]:
```suggestion
        const navigationScreen = success ? 'Login' : 'Counter';
```

File: modules/user/client-react/containers/ActivateUser.jsx [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+        const webActions = success ? history.push('/login/') : history.push('/');
```
Comment (alekseyzadorozhniy) [inline modules/user/client-react/containers/ActivateUser.jsx]:
```suggestion
        const navigationUrl = success ? '/login' : '/';
```

File: modules/user/client-react/containers/ActivateUser.jsx [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+        return navigation ? mobileActions : webActions;
```
Comment (alekseyzadorozhniy) [inline modules/user/client-react/containers/ActivateUser.jsx]:
```suggestion
        navigation.navigate(navigationScreen);
        history.push(navigationUrl);
```

File: modules/user/client-react/components/RegisterView.native.jsx [hunk 14]
```diff
@@ -1,14 +1,35 @@
...
-    t: PropTypes.func
+    t: PropTypes.func,
+    navigation: PropTypes.object
+  };
+
+  async componentDidMount() {
+    Linking.addEventListener('url', this.handleUrl);
+    if (Platform.OS === 'ios') {
+      const url = await Linking.getInitialURL();
+      if (url.includes('/confirmation/')) {
```
Comment (evgenijolshanskij) [inline modules/user/client-react/components/RegisterView.native.jsx]: Remove the logic duplication.

File: modules/user/server-ts/schema.graphql [hunk 15]
```diff
@@ -60,6 +60,9 @@ type UserPayload {
   user: User
   errors: [FieldError!]
 }
+type ActivateUserResponse {
```
Comment (evgenijolshanskij) [inline modules/user/server-ts/schema.graphql]: Is there any other returning type within the App with the prefix `Response`?
